### ISSUE TITLE
Publishing via a lua "location.capture()" subrequest randomly hangs or succeeds

### ISSUE THREAD

**Issue by nkuehn** [slact/nchan#429]
I'm trying to set up a lua-scripted nchan to customize the HTTP interface without having to run a less scalable application server in parallel for the minimal business logic I need to customize the publishing API (e.g. return different logic of HTTP status responses than the default 202/201; e.g. read channel ID from the message body payload etc). 

I have the problem that calling the subrequest to the publisher endpoint via a content_by_lua_block  with ngx.location.capture("/pub", ...)  fails randomly - more specifically it hangs and does not return.   When directly calling /pub it's all fine and when  calling a non-nchan location via subrequest it's fine, too. 

In the minimal setup (see below) the first POST after server start usually succeeds but then it gets random.  In my full, more complex setup it's always hanging. 

To reproduce I have set up a Docker container based on debian that downloads the nchan master and builds nchan on openresty  using your own  '/dev/rebuild.sh' build script (was  a bit of a hassle to get all the dependencies right but this way you can do a 100% exact reproduction). 

The build and the minimal site configuration is in this gist:  https://gist.github.com/nkuehn/d86313556caac4591683269c97d4fecb  

key fragment of the site conf:
```conf
location /pub-via-lua {
		# predictable body handling with fixed body buffers (max message size is known 256k):
		client_max_body_size 256k;
                client_body_buffer_size 256k;
		content_by_lua_block {
			ngx.req.read_body()  -- explicitly read the req body
			local bodydata = ngx.req.get_body_data()
			local res = ngx.location.capture("/pub-ctmessage-post", { method = ngx.HTTP_POST, body = bodydata })
			ngx.print(res.body)
			return -- implies status 200
		}
	}

	location /pub {
		nchan_publisher;
		nchan_channel_id test;
		nchan_channel_group testgroup;
		nchan_eventsource_event testmessage;
		nchan_channel_events_channel_id metapublish;
		access_log /dev/stdout messagepost;         
}
```
(FYI: this is the followup to my question in #422 )

**Comment by slact** [comment]
No time this week, but i'll be investigating this next Monday.

**Comment by nkuehn** [comment]
Thank you so much!  I have no hurry, the effort has no time constraints on my side - I'm "just" trying to prove the concept works.

**Comment by develephant** [comment]
Hi,

I'm loving Nchan, but I too am having a similar issue with the subrequest hanging, but only specific to using the redis server backend. If I remove the redis backend directives the subrequest works fine.

Interestingly, while testing with Postman, the message does get published (I can see it when I check the /sub endpoint), and when I cancel the "hanging" request (via Postman) the result of the subrequest is captured in the log file. So, for whatever reason, the publish endpoint is not returning the result to the initial subrequest and instead hangs.

**nginx.conf** (fragment)

```
nchan_redis_url redis://127.0.0.1:6379;

location ~ /sub/(\w+)$ {
  nchan_subscriber intervalpoll;
  nchan_subscriber_first_message newest;
  nchan_channel_id $1;
  nchan_use_redis on;
}

location ~ /pub/(\w+)$ {
  internal;
  nchan_publisher http;
  nchan_channel_id $1;
  nchan_use_redis on;
}

location = /test {
  content_by_lua_file /usr/local/openresty/lualib/respond.lua;
}
```

**respond.lua**

```
ngx.req.discard_body()

local res, err = ngx.location.capture( '/pub/game', { method = ngx.HTTP_POST,
  body = 'Hello, World'
})

ngx.log(ngx.ERR, res.status) -- this shows up in the log (202) once I cancel the request

ngx.print(res.status) -- this never fires, even after canceling the request
```

Thanks again for Nchan! I can move on without using redis, but it would be nice to have the persistence and scalability.

-dev
